### PR TITLE
clean code of the undo button

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
I found the code which deal with the "undo" button quite ugly. I hope you'll agree with me that this new version is more clear. I removed a lot of code duplication.
Each commit are supposed to be easy to follow, it should be clear nothing is done, with one exception,  https://github.com/ankidroid/Anki-Android/commit/0c37624b5a0d888e573aa8977c80c2ac4ff29e1f is more complex, so the commit message explain why actually, this simplification is valid.

My ultimate goal is to deal with https://github.com/ankidroid/Anki-Android/issues/6119 with a solution which allows the undo button to sometime change its color. I've learned my lesson, I first submit code simplification, and only when it has been accepted do I submit the code which does actual bug correction.
(Normally, the bug will become really easy to correct now that the code is simpler)

### CODE REVIEW

**Comment by Arthur-Milchior** [comment]
Now that https://github.com/ankidroid/Anki-Android/pull/6227 is merged, I added a last commit which change the color of the undo button if and only if it is disabled and furthermore we expect it to be slow. 
I'll remove this commit if you want and prefer to keep the previous behaviour.

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java [hunk 1]
```diff
@@ -431,21 +431,29 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
         } else {
-            menu.findItem(R.id.action_undo).setIcon(R.drawable.ic_undo_white_24dp);
-            menu.findItem(R.id.action_undo).setEnabled(false).getIcon().setAlpha(Themes.ALPHA_ICON_DISABLED_LIGHT);
+            // We can arrive here even if `mShowWhiteboard &&
+            // mWhiteboard != null` if no stroke had ever been made
+            undoIcon = R.drawable.ic_undo_white_24dp;
+            undoEnabled = (colIsOpen() && getCol().undoAvailable());
+        }
+        int alpha;
+        if (getControlBlocked() == ControlBlock.SLOW || !undoEnabled) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:448]: Could you change the title of the PR if there's a functional change mixed in

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java [hunk 2]
```diff
@@ -431,21 +431,29 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
-            menu.findItem(R.id.action_undo).setEnabled(true).getIcon().setAlpha(Themes.ALPHA_ICON_ENABLED_LIGHT);
-        } else if (mShowWhiteboard && mWhiteboard != null && mWhiteboard.isUndoModeActive()) {
-            // Whiteboard undo queue empty, but user has added strokes to it for current card. Disable undo button.
-            menu.findItem(R.id.action_undo).setIcon(R.drawable.ic_eraser_variant_white_24dp);
-            menu.findItem(R.id.action_undo).setEnabled(false).getIcon().setAlpha(Themes.ALPHA_ICON_DISABLED_LIGHT);
-        } else if (colIsOpen() && getCol().undoAvailable()) {
-            menu.findItem(R.id.action_undo).setIcon(R.drawable.ic_undo_white_24dp);
-            menu.findItem(R.id.action_undo).setEnabled(true).getIcon().setAlpha(Themes.ALPHA_ICON_ENABLED_LIGHT);
+        // Undo button
+        int undoIcon;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:435]: Nit: `@DrawableRes` if this needs to exist

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java [hunk 3]
```diff
@@ -218,6 +218,11 @@ public int undoSize() {
         return mUndo.size();
     }
 
+    /** @return Whethere there are strokes to undo */
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java:221]:
```suggestion
    /** @return Whether there are strokes to undo */
```

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java [hunk 4]
```diff
@@ -431,21 +431,29 @@ public boolean onCreateOptionsMenu(Menu menu) {
...
+        boolean undoEnabled;
```
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:436]: "these" ? I don't know what you are referring to.
I don't believe that there is any interest in having undoEnabled as a member of the class. It's quite easy right now to get this information from the undoEmpty and undoAvailable function. If we wanted to keep this information as a method, the code change would get bigger as it means we would need to ensure that this variable is changed each time the collection/whiteboard undo list is changed

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java [hunk 5]
```diff
@@ -402,5 +407,9 @@ public void apply() {
             }
             invalidate();
         }
+
+        public boolean empty() {
```
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java:411]: I wanted to stay consistent with the method name of Stack. After all, UndoStack is essentially a Stack, so keeping name as consistent as possible with the Stack class seems better. After all, that is already what occurs with add, clear, size and pop.

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:448]: Sorry, I thought I already did it. My bad

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:435]: Thanks. I didn't know this annotation

File: AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/Whiteboard.java:411]: So I currently stick with my choice here, unless a second person share your preference

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:436]: Are there other variables you want to move to the class level ?

**Comment by Arthur-Milchior** [comment]
> A few nits. Happy as long as there's confirmation this has been tested against the #6119 regression

That was I tried to explain in my second comment. I tried to reproduce the regression and can not. Which is not really surprising, the only difference occurs if there is a "slow" operation, e.g. suspend/bury/cancel, and those operation empties the whiteboard stack anyway

**Comment by Arthur-Milchior** [comment]
Correction pushed

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:436]: Disregard this, the spacing inside GitHub implied that these were member variables.

**Comment by david-allison-1** [comment]
Needs a `import androidx.annotation.DrawableRes;`, then LGTM

**Comment by Arthur-Milchior** [comment]
Done

**Pull request merged**